File: scripts/lib/ssot_paths.py
#!/usr/bin/env python3
#-*- coding: utf-8 -*-

"""
ssot_paths.py — SSOT Path Bridge for scripts/.

Thin re-export from mas_mcp.logic.ssot_manifest so that scripts/ modules
never hardcode SSOT filenames. If a filename changes, ssot_manifest.py
changes first, and this bridge propagates it everywhere.

Usage in any script:
    from scripts.lib.ssot_paths import SSOT_POINTER, SSOT_HOLDER, resolve_ssot_paths

    root = find_repo_root()
    paths = resolve_ssot_paths(root)
    ssot_content = paths.pointer.read_text(encoding="utf-8")
"""

from __future__ import annotations

import sys
from dataclasses import dataclass
from pathlib import Path

# Ensure repo root is importable (scripts may be run from anywhere)
_REPO_ROOT = Path(__file__).resolve().parent.parent.parent
if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

try:
    from mas_mcp.logic.ssot_manifest import (  # noqa: E402
        SSOT_HOLDER_RELPATH,
        SSOT_POINTER_RELPATH,
        SSOT_PROTO_RELPATH,
    )
except ImportError:
    import warnings
    warnings.warn(
        "mas_mcp.logic.ssot_manifest not importable — using hardcoded fallback paths. "
        "Install the package or run from the repo root.",
        DeprecationWarning,
        stacklevel=2,
    )
    SSOT_HOLDER_RELPATH = ".github/copilot-instructions.archive.md"
    SSOT_POINTER_RELPATH = ".github/copilot-instructions.md"
    SSOT_PROTO_RELPATH = ".github/copilot-instructions-copy.md"

# ─── Re-exported constants (relative paths as strings) ──────────────────────
SSOT_HOLDER = SSOT_HOLDER_RELPATH
SSOT_POINTER = SSOT_POINTER_RELPATH
SSOT_PROTO = SSOT_PROTO_RELPATH


@dataclass(frozen=True)
class SSOTPaths:
    """Resolved absolute SSOT paths for a given repo root."""
    holder: Path
    pointer: Path
    proto: Path


def resolve_ssot_paths(repo_root: Path) -> SSOTPaths:
    """Resolve canonical SSOT paths from a repo root."""
    return SSOTPaths(
        holder=repo_root / SSOT_HOLDER_RELPATH,
        pointer=repo_root / SSOT_POINTER_RELPATH,
        proto=repo_root / SSOT_PROTO_RELPATH,
    )

File: scripts/lib/ssot-paths.ts
#!/usr/bin/env bun

// @SID: BRIDGE_SSOT_PATHS_TS_V1
// @Shabti:        Library Module
// @Shabti-Desc:   SSOT Cascade Bridge for TypeScript — re-exporting from ssot_paths.ts to avoid hardcoding SSOT paths in TS modules.
// @Type:          INFRASTRUCTURE
// @Spectrum:      WHITE
// @Zone:          THE GARDEN
// @Ogdoad:        (Standalone)

/**
* SSOT Cascade Bridge — TypeScript
*
* Mirrors the Python cascade (ssot_manifest.py → ssot_paths.py) for TS consumers.
* All SSOT path references in TypeScript MUST import from here, never hardcode.
 *
 * Canonical source: scripts/lib/ssot-paths.ts
 *
 * If the SSOT filename changes, update THIS file (and the Python manifest).
 * Downstream TS consumers resolve through these exports.
 */

import { existsSync } from "fs";
import { resolve as pathResolve } from "path";

/**
 * The primary, file-first-authority SSOT canon — the frozen monolithic
 * Codex-Brahmanica-Perfectus from which every active instruction file is
 * derived. This is THE single anchor every surface resolves to
 * (canonized in place 2026-05-31). It is the artifact that created everything
 * else; surfaces that reference "the SSOT" mean THIS file.
 */
export const SSOT_CANON = ".chthonic/SSOT.md" as const;

/**
 * Generated downstream MIRROR of SSOT_CANON, kept under this clunky-but-stable
 * name so existing repo scripts that reference it do not break (no-delete).
 * NOT the primary source — resolve SSOT_CANON first; fall back here only if the
 * canon is absent.
 */
export const SSOT_HOLDER = ".github/copilot-instructions.archive.md" as const;

/** The 85-line routing pointer. */
export const SSOT_POINTER = ".github/copilot-instructions.md" as const;

/** The historizing fork (pre-freeze snapshot). */
export const SSOT_PROTO = ".github/copilot-instructions-copy.md" as const;

/**
 * Resolve an SSOT path relative to a given root directory.
 */
export function resolveSsotPath(
  root: string,
  which: "canon" | "holder" | "pointer" | "proto" = "pointer",
): string {
  const map = { canon: SSOT_CANON, holder: SSOT_HOLDER, pointer: SSOT_POINTER, proto: SSOT_PROTO };
  // Use platform-agnostic join
  return `${root}/${map[which]}`.replace(/\\/g, "/");
}

/**
 * Assert the SSOT holder exists at the expected path.
 * Throws with a descriptive message if absent.
 */
export function assertSsotExists(root: string): void {
  const holderPath = pathResolve(root, SSOT_HOLDER);
  if (!existsSync(holderPath)) {
    throw new Error(
      `SSOT holder not found at expected path: ${holderPath}\n` +
      `Ensure repository root is correct (got: ${root})`
    );
  }
}
